core/retries/src/main/java/software/amazon/awssdk/retries/internal/LegacyRetryStrategyImpl.java

```diff
@@ -0,0 +1,385 @@
+        DefaultRetryToken token = asStandardRetryToken(request.token());
+        if (acquireResponse.acquisitionFailed()) {
+            Throwable failure = request.failure();
+            DefaultRetryToken refreshedToken =
+                token.toBuilder()
+                     .capacityRemaining(acquireResponse.capacityRemaining())
+                     .capacityAcquired(acquireResponse.capacityAcquired())
+                     .state(DefaultRetryToken.TokenState.TOKEN_ACQUISITION_FAILED)
+                     .addFailure(failure)
+                     .build();
+            String message = acquisitionFailedMessage(acquireResponse);
+            LOG.error(() -> message, failure);
+            throw new TokenAcquisitionFailedException(message, refreshedToken, failure);
+        }
+    }
+
+    private void throwOnMaxAttemptsReached(RefreshRetryTokenRequest request, AcquireResponse acquireResponse) {
+        DefaultRetryToken token = asStandardRetryToken(request.token());
+        if (maxAttemptsReached(token)) {
+            Throwable failure = request.failure();
+            DefaultRetryToken refreshedToken =
+                token.toBuilder()
+                     .capacityRemaining(acquireResponse.capacityRemaining())
+                     .capacityAcquired(acquireResponse.capacityAcquired())
+                     .state(DefaultRetryToken.TokenState.MAX_RETRIES_REACHED)
+                     .addFailure(failure)
+                     .build();
+            String message = maxAttemptsReachedMessage(refreshedToken);
+            LOG.error(() -> message, failure);
+            throw new TokenAcquisitionFailedException(message, refreshedToken, failure);
+        }
+    }
+
+    private void throwOnNonRetryableException(RefreshRetryTokenRequest request, AcquireResponse acquireResponse) {
+        DefaultRetryToken token = asStandardRetryToken(request.token());
+        Throwable failure = request.failure();
+        if (isNonRetryableException(request)) {
+            String message = nonRetryableExceptionMessage(token);
+            LOG.error(() -> message, failure);
+            DefaultRetryToken refreshedToken =
+                token.toBuilder()
+                     .capacityRemaining(acquireResponse.capacityRemaining())
+                     .capacityAcquired(acquireResponse.capacityAcquired())
+                     .state(DefaultRetryToken.TokenState.NON_RETRYABLE_EXCEPTION)
+                     .addFailure(failure)
+                     .build();
+            throw new TokenAcquisitionFailedException(message, refreshedToken, failure);
+        }
+        int attempt = token.attempt();
+        LOG.warn(() -> String.format("Request attempt %d encountered retryable failure.", attempt), failure);
+    }
+
+    private String nonRetryableExceptionMessage(DefaultRetryToken token) {
+        return String.format("Request attempt %d encountered non-retryable failure", token.attempt());
```

Comment:
String.format is not cheap, let's only invoke it when needed. 

```
LOG.error(() -> nonRetryableExceptionMessage(token), failure);
```

instead of 
```
String message = nonRetryableExceptionMessage(token)
LOG.error(() -> message, failure);
```